examples/recording_service/leveldb-integration/cpp/LevelDbReader.cxx

```diff
@@ -0,0 +1,637 @@
+     * We wrap it inside a smart pointer so we don't have to worry about
+     * cleanup.
+     */
+    std::shared_ptr<DDS_TypeObjectFactory> type_factory(
+            DDS_TypeObjectFactory_new(),
+            DDS_TypeObjectFactory_delete);
+    UserDataKey current_key;
+    const int64_t timestamp_limit =
+            to_nanosec_timestamp(selector.time_range_end());
+    const int32_t max_samples = selector.max_samples();
+    int32_t num_samples_read = 0;
+    bool next_sample_valid = db_iterator_->Valid();
+    while (next_sample_valid
+            && current_timestamp_ <= timestamp_limit
+            && (max_samples < 0 ? true : (num_samples_read < max_samples))) {
+        ReducedDCPSPublication reduced_sample =
+                slice_to_user_type<ReducedDCPSPublication>(
+                        db_iterator_->value(),
+                        value_buffer_);
+        std::shared_ptr<rti::routing::StreamInfo> stream_info =
+                std::make_shared<rti::routing::StreamInfo>("", "");
+        if (reduced_sample.valid_data()) {
+            stream_info->stream_name(reduced_sample.topic_name());
+            stream_info->type_info().type_name(reduced_sample.type_name());
+            stream_info->type_info().type_representation_kind(
+                    rti::routing::TypeRepresentationKind::DYNAMIC_TYPE);
+
+            DDS_TypeObject *type_obj =
+                    DDS_TypeObjectFactory_create_typeobject_from_serialize_buffer(
+                            type_factory.get(),
+                            (const char *) reduced_sample.type().data(),
+                            reduced_sample.type().size());
+            if (type_obj == nullptr) {
+                std::stringstream log_stream;
+                log_stream << "Failed to create type-object from serialized "
+                        "buffer in DB (will be skipped)" << std::endl;
+                log_stream << "    Stream name = "
+                        << reduced_sample.topic_name() << std::endl;
+                std::cerr << log_stream.str();
+                continue;
+            }
+            std::shared_ptr<DDS_TypeObject> type_ptr(
+                    type_obj,
+                    TypeObjectDeleter(type_factory.get()));
+            DDS_TypeCode *type_code =
+                    DDS_TypeObject_convert_to_typecode(type_obj);
+            if (type_code == nullptr) {
+                std::stringstream log_stream;
+                log_stream << "Failed to create type-code from type-object "
+                        "(will be skipped)" << std::endl;
+                log_stream << "    Stream name = "
+                        << reduced_sample.topic_name() << std::endl;
+                std::cerr << log_stream.str();
+                continue;
+            }
+            stream_info->type_info().type_representation(type_code);
+        }
+        loaned_stream_infos_.resize(num_samples_read + 1);
+        loaned_stream_infos_[num_samples_read] = stream_info;
```

Comment:
this making an assignment of the stream_info shared ptr correcT?

Reply:
Yes, that is the case.